File: 프로그래머스/3/92344. 파괴되지 않은 건물/파괴되지 않은 건물.py
def solution(board, skill):
    r = len(board)
    c = len(board[0])

    prefix_sum = [[0] * (c+2) for _ in range(r+2)]

    for i in range(len(skill)) :
        ty, r1,c1,r2,c2,degree = skill[i]
        r1 +=1; r2+=1; c1+=1; c2+=1
        if ty == 1 :
            degree *= -1
        prefix_sum[r1][c1] += degree
        prefix_sum[r2+1][c2+1] += degree
        prefix_sum[r1][c2+1] -= degree
        prefix_sum[r2+1][c1] -= degree
        

#     for i in range(len(prefix_sum)) :
#         for j in range(1,len(prefix_sum[i])) :
#             prefix_sum[i][j] = prefix_sum[i][j-1] + prefix_sum[i][j]

#     for j in range(len(prefix_sum[0])) :
#         for i in range(1,len(prefix_sum)) :
#             prefix_sum[i][j] = prefix_sum[i-1][j] + prefix_sum[i][j]
    
    for i in range(1,r+1) :
        for j in range(1,c+1):
            prefix_sum[i][j] += prefix_sum[i-1][j] + prefix_sum[i][j-1] - prefix_sum[i-1][j-1]

    answer = 0
    for i in range(r) :
        for j in range(c) :
            board[i][j] += prefix_sum[i+1][j+1]
            if board[i][j] > 0 : answer += 1

    return answer

File: 프로그래머스/3/92344. 파괴되지 않은 건물/파괴되지 않은 건물.java
class Solution {
    public int solution(int[][] board, int[][] skill) {
        int n = board.length;
        int m = board[0].length;
        int[][] imos = new int[n + 1][m + 1];
        
        // 스킬 적용
        for (int[] s : skill) {
            int type = s[0], r1 = s[1], c1 = s[2], r2 = s[3], c2 = s[4], degree = s[5];
            int value = type == 1 ? -degree : degree;
            
            imos[r1][c1] += value;
            imos[r2 + 1][c1] -= value;
            imos[r1][c2 + 1] -= value;
            imos[r2 + 1][c2 + 1] += value;
        }
        
        
        
        
        // 누적합 계산
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                if (i > 0) imos[i][j] += imos[i - 1][j];
                if (j > 0) imos[i][j] += imos[i][j - 1];
                if (i > 0 && j > 0) imos[i][j] -= imos[i - 1][j - 1];
                
                board[i][j] += imos[i][j];
            }
            
        }
        
        // 파괴되지 않은 건물 수 계산
        int answer = 0;
        for (int i = 0; i < n; i++) {
            for (int j = 0; j < m; j++) {
                if (board[i][j] > 0) answer++;
            }
        }
        
        return answer;
    }
    
    private static void print(int n, int m, int[][] board) {
        for(int i=0; i<n; i++) {
            for(int j=0; j<m; j++) {
                System.out.print(board[i][j] + " ");
            }
            System.out.println();
        }
        System.out.println();
    }
}